source/agora/consensus/EnrollmentManager.d

```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
         return true;
     }
 
+    /***************************************************************************
+
+        Check if a pre-image exists
+
+        Params:
+            enroll_key = The key for the enrollment in which the pre-image is
+                contained.
+            height = The block height of the preimage to check existence
+
+        Returns:
+            true if the pre-image exists
+
+    ***************************************************************************/
+
+    public bool hasPreimage (const ref Hash enroll_key, ulong height) @trusted
+    {
+        auto results = this.db.execute("SELECT preimage from validator_set" ~
+                    " WHERE key = ?", enroll_key.toString());
+        if (!results.empty && results.oneValue!(byte[]).length != 0)
+        {
+            PreimageInfo loaded_image =
+                results.oneValue!(ubyte[]).deserializeFull!(PreimageInfo);
+            if (height <= loaded_image.height)
+                return true;
+        }
+        return false;
+    }
+
+    /***************************************************************************
+
+        Add a pre-image information to a validator data
+
+        Params:
+            preimage = the pre-image information to add
+
+        Returns:
+            true if the pre-image information has been added to the validator
+
+    ***************************************************************************/
+
+    public bool addPreimage (const ref PreimageInfo preimage) @safe nothrow
+    {
+        static ubyte[] buffer;
```

Comment:
This needs to be moved right before the `buffer.length = 0;` statement. Declarations should be placed close to where they're used the first time, which is not necessarily at the start of the function.